metals/src/main/scala/scala/meta/internal/metals/Compilations.scala

```diff
@@ -0,0 +1,121 @@
+package scala.meta.internal.metals
+import ch.epfl.scala.{bsp4j => b}
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+
+final class Compilations(
+    buildTargets: BuildTargets,
+    classes: BuildTargetClasses,
+    workspace: () => AbsolutePath,
+    buildServer: () => Option[BuildServerConnection]
+)(implicit ec: ExecutionContext) {
+
+  // we are maintaining a separate queue for cascade compilation since those must happen ASAP
+  private val compilationBatch = new CompilationBatch
+  private val cascadeBatch = new CascadeCompilationBatch
+
+  private val isCompiling = TrieMap.empty[b.BuildTargetIdentifier, Boolean]
+  private var lastCompile: collection.Set[b.BuildTargetIdentifier] = Set.empty
+
+  def currentlyCompiling: Iterable[b.BuildTargetIdentifier] = isCompiling.keys
+  def currentlyCompiling(buildTarget: b.BuildTargetIdentifier): Boolean =
+    isCompiling.contains(buildTarget)
+
+  def previouslyCompiled: Iterable[b.BuildTargetIdentifier] = lastCompile
+  def previouslyCompiled(buildTarget: b.BuildTargetIdentifier): Boolean =
+    lastCompile.contains(buildTarget)
+
+  def compileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] =
+    compilationBatch.enqueue(paths)
+
+  def cascadeCompileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] =
+    cascadeBatch.enqueue(paths)
+
+  def cancel(): Unit = {
+    compilationBatch.cancel()
+    cascadeBatch.cancel()
+  }
+
+  private abstract class Batch extends Cancelable {
```

Comment:
Can we try to remove this class? This is a heavy abstraction IMO for a simple `Seq[BuildTargetIdentifier] => Seq[BuildTargetIdentifier]` function. 

I would prefer to keep
```scala
private val compile = new BatchedFunction[b.BuildTargetIdentifier, b.CompileResult](compile)
private val cascade = new BatchedFunction[b.BuildTargetIdentifier, b.CompileResult](compile)
```
and then expand the targets at the call-site
```scala
 def compileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] =
    compile(expand(paths))

  def cascadeCompileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] =
    cascade(expand(paths).flatMap(inverseDependencies).distinct)
```
Benefits of this approach:
- no need to come up with slightly different but new names for existing abstactions: `Batch` instead of `BatchedFunction`, `Batch.enqueue` instead of `BatchedFunction.apply`
- uses no abstract methods like `Batch.resolve`, which are tricky to reason about in my experience